client/trino-client/src/main/java/io/trino/client/auth/external/SharedToken.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.auth.external;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+class SharedToken
+        implements KnownToken
+{
+    private static final ConcurrentHashMap<String, Token> sharedTokens = new ConcurrentHashMap<>();
+
+    private final String userName;
+
+    SharedToken(String userName)
+    {
+        this.userName = requireNonNull(userName, "userName is null");
+    }
+
+    @Override
+    public Optional<Token> getToken()
+    {
+        return Optional.ofNullable(sharedTokens.get(userName));
+    }
+
+    @Override
+    public void setupToken(Supplier<Token> tokenSource)
```

Comment:
I would make this method simply synchronized and same for `getToken`. Then you can use plain map.